src/metaopt/worker/experiment.py

```diff
@@ -0,0 +1,354 @@
+
+        return selected_trial
+
+    def push_completed_trial(self, trial):
+        """Inform database about an evaluated `trial` with results.
+
+        :param trial: Corresponds to a successful evaluation of a particular run.
+        :type trial: `Trial`
+
+        .. note:: Change status from *reserved* to *completed*.
+        """
+        trial.status = 'completed'
+        self._db.write('trials', dict(trial), query={'_id': trial.id})
+
+    def register_trials(self, trials):
+        """Inform database about *new* suggested trial with specific parameter
+        values. Each of them correspond to a different possible run.
+
+        :type trials: list of `Trial`
+        """
+        for trial in trials:
+            trial.status = 'new'
+        self._db.write('trials', list(map(dict, trials)))
+
+    @property
+    def is_done(self):
+        """Count how many trials have been completed and compare with `max_trials`.
+
+        .. note:: To be used as a terminating condition in a ``Worker``.
+        """
+        query = dict(
+            exp_name=self.name,
+            user=self.metadata['user'],
+            status='completed'
+            )
+        num_completed_trials = len(self._db.read('trials', query, {'_id': 1}))
+        if num_completed_trials >= self.max_trials:
+            return True
+        return False
+
+    @property
+    def configuration(self):
+        """Return a copy of an `Experiment` configuration as a dictionary."""
+        if self._init_done:
+            return self._db.read('experiments', {'_id': self._id})[0]
+
+        config = dict()
+        for attrname in self.__slots__:
+            if not attrname.startswith('_'):
+                config[attrname] = getattr(self, attrname)
+        # Reason for deepcopy is that some attributes are dictionaries
+        # themselves, we don't want to accidentally change the state of this
+        # object from a getter.
+        return copy.deepcopy(config)
+
+    def configure(self, config):
+        """Set `Experiment` by overwriting current attributes.
+
+        If `Experiment` was already set and an overwrite is needed, a *fork*
+        is advised with a different :attr:`name` for this particular configuration.
+
+        .. note:: Calling this property is necessary for an experiment's
+           initialization process to be considered as done. But it can be called
+           only once.
+        """
+        if self._init_done:
+            raise RuntimeError("Configuration is done; cannot reset an Experiment.")
+
+        # If status is None in this object, then database did not hit a config
+        # with same (name, user's name) pair. Everything depends on the user's
+        # moptconfig to set.
+        if self.status is None:
+            if config['name'] != self.name or \
+                    config['metadata']['user'] != self.metadata['user'] or \
+                    config['metadata']['datetime'] != self.metadata['datetime']:
```

Comment:
This check is for my well-being and developer check. Reminds that external configuration must be based on `Experiment.configuration` first, to draw the inferred `metadata.user` and `metadata.datetime`. `resolve_config` could happen semantically inside  `Experiment`, but I implemented it outside beforehand so that I do not delegate too much in `Experiment`. So finally, `Experiment` infers some implied (datetime and user) things (which now I have the feeling that this is not good (at least for `user`, I can give arguments in an issue later)) and sets up whatever is needed for a working experiment (checks and instantiates if needed `refers` and `algorithms`).

This exception here shows that it is meant to be used somehow like this:
```
my_config = get_my_config()
exp = Experiment(my_config['name'])
inferred_config = exp.configuration
final_config = merge_configs(my_config, inferred_config)
exp.configure(final_config)
```

It says that `datetime` and `user` cannot be effectively set from user's config. Keep them inferred.

Reply:
I don't understand why it could not be like that

```python
my_config = get_my_config()
exp = Experiment(my_config['name'])
exp.configure(my_config)
```

merge_configs should be something happening inside Experiment in my opinion. Experiment is already trying to update it's config with exp.configure so I can hardly see why it could not do merge_configs at that point rather than outside of it beforehand. I understand your point about delegation and I believe merge_configs is something that **should** be delegated to Experiment.

Reply:
in `cli.py`, line 69-70:
```
expconfig = resolve_config.merge_mopt_config(expconfig, dbconfig,
                                             cmdconfig, cmdargs)
```
This is because of the hierarchical configuration. It's too much to get all of these in `Experiment`.

Reply:
I agree this global merge should not happen in ```Experiment```. However I don't see why we would need to fetch the config from db in order to do the merge. Can't we merge ```dbconfig```, ```cmdconfig``` and ```cmdargs``` along with user's provided ```expconfig``` without fetching the db and then pass this to ```exp.configure()```? The configuration passed to experiment could be incomplete but that's not a problem if ```configure()``` only overwrite attributes and do not expect to overwrite **all** of them. Furthermore, changes from ```expconfig``` might trigger a _forking event_ which so far is planned to be handled by Experiment, so it would make sense for the _final_ merge to occur in Experiment.

In the same spirit, maybe ```cli.py``` does not need to know or maybe it **should not** know about internal experiment configuration fetched from db. ```cli.py``` would be just fetching config provided by user and pass it to ```Experiment``` which do whatever it wants with it. From my point of view, we are delegating _db-merge_ to ```cli.py``` to avoid delegating global merge to Experiment. Both could be avoided.

Reply:
The thing is that how I have set it up currently the configurations have the following precedence: `expconfig` (defaults) < `dbconfig` (fetched from db if it has hit) < `cmdconfig` (file provided by cmd args) < `cmdargs` (for things in cmd args). It cannot be done inside the `Experiment` object without passing logic inside it. So in order not to overload it, there is the `resolve_config` module with the function. `cli.py` just organizes. Information passed to `exp.configure(...)` should be final resolved will and configuration. Nonetheless, `cli.py` organizes the instantiation of the modules and controller objects, I do not see how it should not have access to some experiment's settings.

Reply:
`dbconfig` is configuration fetched from the database.

Reply:
I think we should discuss this configuration structure a bit more but let's move the discussion to an issue and leave it like that for this PR so that you can move on to Worker/Producer/Consumer.